### PR TITLE
[Fix #53581] Fix error when building devcontainer

### LINKED ISSUE #53581: [rails-dev] devcontainer based development appears broken due to permissions issue during `apt-get`
> [!NOTE]  
> This issue could _not_ be related to Rails but I've been unable to track it down. Curious if anyone else has hit it. If it's deemed to not be Rails related, we can close this issue.

### Steps to reproduce

Set up Rails VSCode devcontainer, pull `main`, do a container rebuild via VSCode actions menu -> fails. ([full log](https://gist.github.com/larouxn/747eab6b79808a0360d9c64e06aad626))

```console
 => ERROR [rails dev_container_auto_added_stage_label  2/19] RUN apt-get   1.3s
------
 > [rails dev_container_auto_added_stage_label  2/19] RUN apt-get update && export DEBIAN_FRONTEND=noninteractive     && apt-get -y install --no-install-recommends         mariadb-client libmariadb-dev         postgresql-client postgresql-contrib libpq-dev         ffmpeg mupdf mupdf-tools libvips-dev poppler-utils         libxml2-dev sqlite3 imagemagick:
1.249 Reading package lists...
1.265 E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)
------
[2024-11-10T10:11:22.225Z] failed to solve: process "/bin/sh -c apt-get update && export DEBIAN_FRONTEND=noninteractive     && apt-get -y install --no-install-recommends         mariadb-client libmariadb-dev         postgresql-client postgresql-contrib libpq-dev         ffmpeg mupdf mupdf-tools libvips-dev poppler-utils         libxml2-dev sqlite3 imagemagick" did not complete successfully: exit code: 100
[2024-11-10T10:11:22.229Z] Stop (33074 ms): Run: docker compose --project-name rails_devcontainer -f /home/larouxn/src/github.com/larouxn/rails/.devcontainer/compose.yaml -f /home/larouxn/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data/docker-compose/docker-compose.devcontainer.build-1731233449154.yml build
[2024-11-10T10:11:22.231Z] Error: Command failed: docker compose --project-name rails_devcontainer -f /home/larouxn/src/github.com/larouxn/rails/.devcontainer/compose.yaml -f /home/larouxn/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data/docker-compose/docker-compose.devcontainer.build-1731233449154.yml build
[2024-11-10T10:11:22.231Z]     at Pm (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:430:525)
[2024-11-10T10:11:22.231Z]     at async ftA (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:430:2476)
[2024-11-10T10:11:22.231Z]     at async htA (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:410:3496)
[2024-11-10T10:11:22.231Z]     at async TtA (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:482:4021)
[2024-11-10T10:11:22.231Z]     at async iB (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:482:4963)
[2024-11-10T10:11:22.231Z]     at async wrA (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:663:203)
[2024-11-10T10:11:22.231Z]     at async DrA (/home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:662:14830)
[2024-11-10T10:11:22.232Z]     at async /home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js:482:1190
[2024-11-10T10:11:22.240Z] Stop (44751 ms): Run: /usr/share/code/code /home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js up --user-data-folder /home/larouxn/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-session-data-folder /tmp/devcontainers-8eed6faa-bd4e-48dc-8274-fd689b1871211731233434623 --workspace-folder /home/larouxn/src/github.com/larouxn/rails --workspace-mount-consistency cached --gpu-availability detect --id-label devcontainer.local_folder=/home/larouxn/src/github.com/larouxn/rails --id-label devcontainer.config_file=/home/larouxn/src/github.com/larouxn/rails/.devcontainer/devcontainer.json --log-level debug --log-format json --config /home/larouxn/src/github.com/larouxn/rails/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --mount type=bind,source=/run/user/1000/wayland-0,target=/tmp/vscode-wayland-2cf7d942-ecba-4d25-a5e9-4172092020f4.sock --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root --include-configuration --include-merged-configuration
[2024-11-10T10:11:22.240Z] Exit code 1
[2024-11-10T10:11:22.249Z] Command failed: /usr/share/code/code /home/larouxn/.vscode/extensions/ms-vscode-remote.remote-containers-0.390.0/dist/spec-node/devContainersSpecCLI.js up --user-data-folder /home/larouxn/.config/Code/User/globalStorage/ms-vscode-remote.remote-containers/data --container-session-data-folder /tmp/devcontainers-8eed6faa-bd4e-48dc-8274-fd689b1871211731233434623 --workspace-folder /home/larouxn/src/github.com/larouxn/rails --workspace-mount-consistency cached --gpu-availability detect --id-label devcontainer.local_folder=/home/larouxn/src/github.com/larouxn/rails --id-label devcontainer.config_file=/home/larouxn/src/github.com/larouxn/rails/.devcontainer/devcontainer.json --log-level debug --log-format json --config /home/larouxn/src/github.com/larouxn/rails/.devcontainer/devcontainer.json --default-user-env-probe loginInteractiveShell --mount type=volume,source=vscode,target=/vscode,external=true --mount type=bind,source=/run/user/1000/wayland-0,target=/tmp/vscode-wayland-2cf7d942-ecba-4d25-a5e9-4172092020f4.sock --skip-post-create --update-remote-user-uid-default on --mount-workspace-git-root --include-configuration --include-merged-configuration
[2024-11-10T10:11:22.249Z] Exit code 1
```

Interestingly, I even tried on an older _used to work a few days ago_ branch and that too failed to build which makes me think this issue could potentially be the VSCode 1.95.2 update or perhaps a silent dev-containers extension update. For what it's worth I have tried the pre-release version of the aforementioned extension as well as rebuilding without cache but neither helped.

Note, I have reproduced this on two separate physical machines as well as a completely fresh Fedora virtual machine.

Screenshot From 2024-11-10 22-20-47

### Expected behavior

1. Install and start Docker
2. Clone `rails/rails`
3. Open repo in VSCode
4. Switch to devcontainer
5. Everything builds properly 

Setup is based on the following, which worked well until today. Last time I used it successfully was probably Thursday.

https://guides.rubyonrails.org/v8.0/contributing_to_ruby_on_rails.html#using-vs-code-remote-containers

### Actual behavior

Container fails to build due to permission error, see above.

### Workaround

Adding `USER root` prior to dependencies step fixes it, above the following.

https://github.com/rails/rails/blob/852d0cd4123463cf215f4b024801b256857295c4/.devcontainer/Dockerfile#L7-L13

### System configuration
**Rails version**: 8.0 (main)

**Ruby version**: 3.3.6

### PR DESCRIPTION
**Pull request by jchilders**
### Motivation / Background

Fixes #53581 

### Detail

When building the Rails devcontainer the `apt-get` step of the Dockerfile was failing with a permissions error. Specifically:

```
> devcontainer build --workspace-folder .
...
=> ERROR [rails dev_container_auto_added_stage_label  2/19] RUN apt-get update && export DEBIAN_FRONTEND=noninteractive     && apt-get -y install --no-install-recommends         mariadb-client libmariadb-dev         postgr  0.1s
0.151 Reading package lists...
0.158 E: List directory /var/lib/apt/lists/partial is missing. - Acquire (13: Permission denied)
```

This PR fixes the permissions error by using `sudo` as appropriate.

The bug this PR fixes was introduced by #53554, but why this is so is not entirely clear.

### Additional information

N/A

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [N/A] Tests are added or updated if you fix a bug or add a feature.
* [N/A] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by zzak** [comment]
I'm not sure providing root permissions is ideal.

> The bug this PR fixes was introduced by https://github.com/rails/rails/pull/53554, but why this is so is not entirely clear.

Could you please investigate why this change introduces breaking behavior? Is it because the image itself is different?

**Comment by jchilders** [comment]
> Is it because the image itself is different?

@zzak I assume so. @matthewd mentions [rails/devcontainer#51](https://github.com/rails/devcontainer/pull/51), but I'm not sure that's related since the breaking change was to Rails proper. I don't see any defined relationship between rails/rails and rails/devcontainer, but I could be wrong.

I'll investigate this a little bit more, but in the meantime I've updated the PR so that the Dockerfile uses `sudo` where needed rather than switching to root.

I would like to reiterate, though, that using `USER root` is (arguably) acceptable here, given that this is limited to local development and a container within it, with no expectation for that container to be used in any other environment.

**Comment by jchilders** [comment]
Ok, finally had some time to look at this. 

tl;dr: It looks like [rails/devcontainer#51](https://github.com/rails/devcontainer/pull/51) is indeed the culprit. Steps to reproduce:

1. Clone `rails/devcontainer` and `cd` into it.
2. From there `cd images/ruby`. This contains the `.devcontainer` directory used to build the Ruby image used by Rails.
3. Build it:

```
devcontainer build --workspace-folder --image-name ruby335:user_vscode .`
```

Since we are building from main, the image will have the `USER vscode` line in the Dockerfile that was introduced in [rails/devcontainer#51](https://github.com/rails/devcontainer/pull/51).

4. Clone `rails/rails` if you haven't already.
5. Edit `.devcontainer/Dockerfile`. Change the `FROM` statement to reference the image we built above:
```
FROM ruby335:user_vscode
```
6. Run the following:

```
devcontainer build --workspace-folder --image-name rails:user_vscode .`
```

It errors out with the same 'Permission denied' error discussed above in this PR.

Rolling back the changes from [rails/devcontainer#51](https://github.com/rails/devcontainer/pull/51) fixes it:

1. Go back to `rails/devcontainer`.
2. Edit `images/ruby/.devcontainer/Dockerfile`
3. Comment out the last line:

```
USER vscode # comment this out
```
4. Build the image again, this time with a different image name:

```
devcontainer build --workspace-folder --image-name ruby335:user_empty .e
```
5. Go back to `rails/rails`.
6. Edit `.devcontainer/Dockerfile` and change the `FROM` statement to:
```
FROM ruby335:user_empty
```
7. Build the rails image again. The error will magically disappear!

```
devcontainer build --workspace-folder --image-name rails:user_empty .`
```

Thoughts on how to proceed with this PR? Should [rails/devcontainer#51](https://github.com/rails/devcontainer/pull/51) be reverted?

**Comment by jchilders** [comment]
@zzak thoughts?
